File: Algorithms/p394_Decode_String.cpp
class Solution {
public:
    string decodeString(string s) {
        string res;
        int len = s.size();
        int j = 0;
        while(1){
            int i = j;
            while(j<len && !(s[j]>='0' && s[j]<='9')){
                ++j;
            }
            res.append(s,i,j-i);
            if(j==len){
                break;
            }
            i = j;
            ++j;
            while(s[j]>='0' && s[j]<='9'){
                ++j;
            }
            int num = stoi(s.substr(i,j-i));
            // s[j]=='['
            int open_brackets = 1;
            ++j;
            i = j;
            while(1){
                if(s[j]=='['){
                    ++open_brackets;
                }else if(s[j]==']'){
                    --open_brackets;
                    if(open_brackets==0){
                        break;
                    }
                }
                ++j;
            }
            string s1 = decodeString(s.substr(i,j-i));
            for(int k=0;k<num;++k){
                res += s1;
            }
            ++j;
        }
        return res;
    }
};

File: Algorithms/p394_Decode_String/p394_Decode_String.py
class Solution(object):
    def decodeString(self, s):
        """
        :type s: str
        :rtype: str
        """
        l = []
        s_l = len(s)
        j = 0
        while True:
            i = j
            while j<s_l and not (ord(s[j])>=ord('0') and ord(s[j])<=ord('9')):
                j += 1
            l.append(s[i:j])
            if j==s_l:
                break
            i = j
            j += 1
            while ord(s[j])>=ord('0') and ord(s[j])<=ord('9'):
                j += 1
            num = int(s[i:j])
            # s[j]=='['
            open_brackets = 1
            j += 1
            i = j
            while True:
                if s[j]=='[':
                    open_brackets += 1
                elif s[j]==']':
                    open_brackets -= 1
                    if open_brackets==0:
                        break;
                j += 1
            l.append(self.decodeString(s[i:j])*num)
            j += 1
        return ''.join(l)
